tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,324 @@
+
+// Access to OnceCell's value is guarded by the semaphore permit
+// and atomic operations on `value_set`, so as long as T itself is Send
+// it's safe to send it to another thread
+unsafe impl<T: Send> Send for OnceCell<T> {}
+
+/// Error returned from the [`OnceCell::set`] method
+///
+/// [`OnceCell::set`]: crate::sync::OnceCell::set
+#[derive(Debug, PartialEq)]
+pub struct AlreadyInitializedError;
+
+impl fmt::Display for AlreadyInitializedError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "AlreadyInitializedError")
+    }
+}
+
+impl Error for AlreadyInitializedError {}
+
+/// Error returned from the [`OnceCell::get`] method
+///
+/// [`OnceCell::get`]: crate::sync::OnceCell::get
+#[derive(Debug, PartialEq)]
+pub struct NotInitializedError;
+
+impl fmt::Display for NotInitializedError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "NotInitializedError")
+    }
+}
+
+impl Error for NotInitializedError {}
+
+/// Allows one to lazily call an async function
+///
+/// # Examples
+/// ```
+/// use tokio::sync::Lazy;
+///
+///
+/// async fn some_computation() -> u32 {
+///     1 + 1
+/// }
+///
+/// static LAZY : Lazy<u32> = Lazy::new(|| Box::pin(some_computation()));
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let result = tokio::spawn(async {
+///         LAZY.get().await
+///     }).await.unwrap();
+///     assert_eq!(*result, 2);
+/// }
+/// ```
+pub struct Lazy<T, F = fn() -> Pin<Box<dyn Future<Output = T> + Send>>> {
+    cell: OnceCell<T>,
+    init: UnsafeCell<Option<F>>,
+}
+
+impl<T, F> Lazy<T, F> {
+    /// Creates a new lazy value with the given initializing
+    /// async function.
+    pub const fn new(init: F) -> Lazy<T, F> {
+        Lazy {
+            cell: OnceCell::new(),
+            init: UnsafeCell::new(Some(init)),
+        }
+    }
+
+    fn get_init(&self) -> &Option<F> {
+        // SAFETY: We only ever give out immutable references to `self.init`
+        unsafe { &*self.init.with(|ptr| ptr) }
+    }
+}
+
+impl<T, F: Fn() -> Pin<Box<dyn Future<Output = T> + Send>>> Lazy<T, F> {
+    /// On first call runs the function stored in Lazy and stores its return value.
+    /// Each subsequent call only returns the stored value.
+    pub async fn get(&self) -> &T {
+        // SAFETY: We only ever give out immutable references
+        let get_f = || match self.get_init() {
+            Some(f) => f(),
+            None => panic!(),
+        };
+
+        self.cell.get_or_init(get_f).await
+    }
+}
+
+impl<T: fmt::Debug, F> fmt::Debug for Lazy<T, F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Lazy")
+            .field("cell", &self.cell)
+            .field("init", &"..")
+            .finish()
+    }
+}
+
+// We never create a `&F` from a `&Lazy<T, F>` so it is fine
+// to not require a `Sync` bound on `F`
+unsafe impl<T, F: Send> Sync for Lazy<T, F> where OnceCell<T>: Sync {}
```

Comment:
This is not true. You do this inside `get`. You can remove both impls, since if you remove the unnecessary `UnsafeCell`, you don't need any unsafe to implement `Lazy`.